src/topology/metric_space/dilation.lean

```diff
@@ -0,0 +1,274 @@
+/-
+Copyright (c) 2022 Hanting Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Dilations of emetric and metric spaces
+Authors: Hanting Zhang
+-/
+import topology.metric_space.antilipschitz
+import data.fun_like.basic
+
+/-!
+# Dilations
+
+We define dilations, i.e., maps between emetric spaces that
+satisfy `edist (f x) (f y) = r * edist x y`.
+
+Defines `ratio f`, which is the ratio of the dilation `f`
+
+Here `r : ℝ≥0`, so we do not exclude the degenerate case of dilations
+which collapse into constant maps. Since there is no `ℝ>0` API defined in mathlib,
+no matter where you choose to exclude r = 0, you will end up having to carry
+it around with you anyways. So statements that do need strict dilations should
+just say `f : dilation α β r` and `hr : r ≠ 0`.
+
+TODO: Introduce dilation equivs. Refactor the isometry API
+to match the `*_hom_class` API below.
+
+Since a lot of elementary properties don't require `eq_of_dist_eq_zero` we start setting up the
+theory for `pseudo_metric_space` and we specialize to `metric_space` when needed.
+-/
+
+noncomputable theory
+
+open function set
+open_locale topological_space ennreal nnreal
+
+section setup
+
+variables (α : Type*) (β : Type*) [pseudo_emetric_space α] [pseudo_emetric_space β]
+
+/-- A dilation is a map that uniformly scales the edistance between any two points.  -/
+structure dilation :=
+(to_fun : α → β)
+(edist_eq' : ∃ r : ℝ≥0, ∀ x y : α, edist (to_fun x) (to_fun y) = r * edist x y)
```

Comment:
Users working with metric spaces may want to work with just `dist` or `nndist` not `edist`. It would make sense to have corresponding `def`s to construct a `dilation` from a `dist_eq` or `nndist_eq` proposition.